Handing in the project for week 15

HenrikeW opened the pull request:
Link to my deployed project: https://henrikes-todo.netlify.app/

My reflections on how this week's project turned out: I think this project was really fun! It was the right level of challenge where I had to think and try different approaches most of the times but never got really stuck. I did the requirements for the red level but would have liked to implement more features if there had been more time. I might go back to this project later on to implement those (most of all a local storage) to be able to actually use the app.

Things I'd like to have clarified or explained in more detail: I used styled components but still struggle with them and would like to learn more about cases where they actually have advantages compared to plain css, like e.g. when passing props. In nearly all components I felt that I could have written the same amount of code using classnames instead but I feel that it's probably because I don't see the whole potential of styled components yet.

MindstormingAB commented on README.md:
```diff
@@ -1,13 +1,53 @@
+- toggle done / not done - button
+- add a task,
+- mark all tasks as done,
+- clear the done tasks and
+- clear the whole list.
+
+The store will keep an array of task-objects and every object has for properties: id, text, done, created where the last one contains a time stamp.
+
+The styling is done with styled components which are stored in three files: one for the header part of the app, one for the central list part and one for the bottom part.
+
+The components' follow a similar structure where 'Header', 'Tasks' and 'Bottom' are the main components with some additional ones nested inside. 
+
+For the 'clear list'-button I added a confirmation dialogue that will open when the button is clicked to make sure the list isn't cleared accidentally. For this I used the package "Sweetalert". 
+
+## Limitations
+
+So far, the site doesn't have a local storage yet and will go back to the default list as soon as you refresh the page. This is something I would have added if I had had more time for the project. Also, I would have liked to implement a priority-function for the tasks and filters to e.g. display only the tasks with highest priority. 
+
+## Compatibility
+
+The app's design is mainly made for mobile use but it works on tablet and desktop as well. I tested it using the dev tools in Chrome, Firefox and Edge where everything works as expected.
+
+## Tech used
+
+- HTML5
+- CSS3
+- JavaScript ES6
+- React
+- Redux
+
+### Additional packages
+
+- styled components
+- moment.js
+- sweetalert
 
 ## View it live
 
-Every project should be deployed somewhere. Be sure to include the link to the deployed project so that the viewer can click around and see what it's all about.
+You can see the site up and running here: https://henrikes-todo.netlify.app/
```
Wow! 🤩

MindstormingAB commented on code/public/index.html:
```diff
@@ -4,6 +4,8 @@
   <head>
     <meta charset="utf-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1" />
+    <link rel="preconnect" href="https://fonts.gstatic.com">
```
What is preconnect? ⁉

MindstormingAB commented on code/src/components/AddTask.js:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  return (
+    <ListItem>
+      <AddForm onSubmit={onAddTask}>
```
I understand that you want to reuse your styled components in different places to keep the code short and consistent but, in this case, you're ending up with a list item outside the list. Maybe you could use more styling in `<AddForm>` and use it as container instead of `<ListItem>`.

HenrikeW replied:
Thanks for the tip! Did that.

MindstormingAB commented on code/src/components/AddTask.js:
```diff
@@ -0,0 +1,38 @@
+    <ListItem>
+      <AddForm onSubmit={onAddTask}>
+        <label htmlFor='newTask'>
+        <img src={require('assets/plus.svg')} alt='' style={{width: '20px'}}/>
```
Mind the indentation and put some useful description in `alt` 😏
What about creating styled component for your image so you can avoid the inline style?

MindstormingAB commented on code/src/components/AddTask.js:
```diff
@@ -0,0 +1,38 @@
+          required
+          />
+        <AddButton 
+          type='submit' 
```
If you use an input as Button, you should add `role= 'button'` for accessibility purpose.

MindstormingAB commented on code/src/components/AddTask.js:
```diff
@@ -0,0 +1,38 @@
+    dispatch(todos.actions.addItem(value));
+    setValue('');
+    event.preventDefault();
+  }
```
Is it a missing semicolon I'm spotting? 🔍

MindstormingAB commented on code/src/components/AddTask.js:
```diff
@@ -0,0 +1,38 @@
+      </AddForm>
+    </ListItem>
+  )
+}
```
More missing semicolons 🙄

MindstormingAB commented on code/src/components/Bottom.js:
```diff
@@ -0,0 +1,57 @@
+  };
+
+  const onClearList = () => {
+    swal({
+      title: 'Are you sure?',
+      text:'Do you want to delete all your todo\'s?',
+      buttons: ['No, cancel!', 'Yes, delete!'],
+      dangerMode: true,
+      closeOnClickOutside: false,
+      className:'alert-window'
+    })
```
Interesting

MindstormingAB commented on code/src/components/Bottom.js:
```diff
@@ -0,0 +1,57 @@
+    <BottomContainer>
+      <BottomButton 
+        onClick={onMarkAllDone}
+        disabled={allTasks.length === 0 || notDoneTasks.length === 0 ? true : false}
+      >Mark all as done
+      </BottomButton>
+      <BottomButton 
+        onClick={onClearDone}
+        disabled={allTasks.length === 0 ? true : false}
+      >Remove done tasks
+      </BottomButton>
+      <BottomButton 
+        onClick={onClearList}
+        disabled={allTasks.length === 0 ? true : false}
```
Well done!

MindstormingAB commented on code/src/components/DoneButton.js:
```diff
@@ -0,0 +1,28 @@
+  if (task.done === true) {
+    return (
+      <DoneButtonContainer onClick={onItemDone}>
+        <img src={require('assets/check.svg')} alt='' style={{width: '20px'}}/>
```
`alt`?

MindstormingAB commented on code/src/components/Summary.js:
```diff
@@ -0,0 +1,14 @@
+      <SummaryItem>{doneTasks.length}/{allTasks.length} done</SummaryItem>
+    </SummaryContainer>
+  )
+}
```
Great code but check if you are missing any semicolon

MindstormingAB commented on code/src/components/TaskList.js:
```diff
@@ -0,0 +1,28 @@
+  if (allTasks.length === 0) {
+    return (
+      <CelebrationContainer>
+        <img src={require('assets/Party.svg')} alt='' style={{width: '45vh'}}/>
```
`alt`?

MindstormingAB commented on code/src/reducers/todos.js:
```diff
@@ -0,0 +1,63 @@
+    addItem: (store, action) => {
+      console.log(action.payload)
+      const newTask = {
+        id: store.tasks.length === 0 ? 0 : Math.max(...store.tasks.map(item => item.id)) + 1,
```
Good job adding this logic to the Math.max id solution

MindstormingAB commented on code/src/reducers/todos.js:
```diff
@@ -0,0 +1,63 @@
+      const itemId = action.payload;
+      const currentItem = store.tasks.find(item => item.id === itemId);
+      
+      currentItem.done === true ? currentItem.done = false : currentItem.done = true;
```
That works but you could have used 
```suggestion
      currentItem.done === !currentItem.done;
```
to toggle the done status

MindstormingAB commented on code/src/reducers/todos.js:
```diff
@@ -0,0 +1,63 @@
+
+        return container
+      });
+      console.log(doneList)
```
console.log is great for testing purposes but don't forget to remove it when you're done 🧐

MindstormingAB commented on code/src/reducers/todos.js:
```diff
@@ -0,0 +1,63 @@
+      const newTaskList = [...store.tasks, newTask];
+      store.tasks = newTaskList;
+    },
+    markAllDone: (store, action) => {
+      const doneList = store.tasks.map((task) => {
+        const container = {}
+
+        container.id = task.id;
+        container.text = task.text;
+        container.done = true;
+
+        return container
+      });
+      console.log(doneList)
+      store.tasks = doneList;
+    },
```
Another (shorter) way to do that:
```suggestion
        markAllDone: (store, action) => {
            store.tasks = store.tasks.map((task) => {
                    task.done = true
                    return task
                })
        },
```

HenrikeW replied:
Thanks for the suggestion. I thought, the longer way was a better way of doing it because of immutability. But maybe I was wrong?

MindstormingAB commented on code/src/reducers/todos.js:
```diff
@@ -0,0 +1,63 @@
+      store.tasks = notDoneList;
+    },
+    clearList: (store, action) => {
+      const emptyList = [];
+
+      store.tasks = emptyList;
```
Or maybe
```suggestion
      store.tasks = [];
```
😉

HenrikeW replied:
Same here. Isn't it better to first create another variable because of immutability reasons?

MindstormingAB commented on code/src/components/DoneButton.js:
```diff
@@ -0,0 +1,28 @@
+        <DoneButtonContainer onClick={onItemDone}>
+          <EmptyCheckbox />
+        </DoneButtonContainer>
+      )
```
So clever! 💯

MindstormingAB commented on code/src/styles/bottomStyles.js:
```diff
@@ -0,0 +1,26 @@
+  padding: 4vh;
+  position: absolute;
+  bottom: 0;
+`
```
Unfortunately, when you have more than 3 todos, this container hides the last todo.

HenrikeW replied:
Solved!

MindstormingAB commented on code/src/styles/bottomStyles.js:
```diff
@@ -0,0 +1,26 @@
+&:disabled {
+  background-color: #ddd6e8;
+  border: 1px solid #28065b;
+}
```
Review indentation 👓

MindstormingAB reviewed (changes requested):
Really great job Henrike!
You demonstrated a good understanding of Redux state and reducers by nailing the blue and red levels requirements. Your app and your code are clean and well designed.
I found several solutions that I'll try to remember for my own projects 🙏 
Before this project can be approved, some changes are required:
- your bottom bar is hiding the last todo when there are more than 3 todos, can you improve that?
- you're using a styled component built on a `li` to wrap your input form which negatively impacts your html semantic.
- your images are missing `alt`
- you're missing some semicolons and the indentation needs to be revised here and there
- remove console.log
Keep up the good work!

HIPPIEKICK commented on code/src/components/Bottom.js:
```diff
@@ -0,0 +1,57 @@
+    <BottomContainer>
+      <BottomButton 
+        onClick={onMarkAllDone}
+        disabled={allTasks.length === 0 || notDoneTasks.length === 0 ? true : false}
```
Nice!

HIPPIEKICK commented on code/src/components/TaskListItem.js:
```diff
@@ -0,0 +1,30 @@
+      />
+      <TextContainer done={task.done}>
+        <p>{task.text}</p>
+        <Timestamp>{moment(task.created).fromNow()}</Timestamp>
```
Nice!

HIPPIEKICK commented:
Great job Henrike - a lot of cool functionality going on here! 💪   Take a look at Sandrine's comments, especially:
- your bottom bar is hiding the last todo when there are more than 3 todos, can you improve that?
- you're missing some semicolons and the indentation needs to be revised here and there
- remove console.log

and you will be approved :)

HenrikeW commented:
Finally, I fixed the footer issue. It's not hiding any content any more. Also, I removed console.logs, added missing semicolons and cleaned up the indentation. As far as I see it, everything should be done now.
